src/apps/intel/intel_app.lua

```diff
@@ -31,25 +34,53 @@ local function firsthole(t)
    end
 end
 
+local provided_counters = {
+   'type', 'dtime', 'mtu', 'speed', 'status', 'promisc', 'macaddr',
+   'rxbytes', 'rxpackets', 'rxmcast', 'rxbcast', 'rxdrop', 'rxerrors',
+   'txbytes', 'txpackets', 'txmcast', 'txbcast', 'txdrop', 'txerrors'
+}
 
 -- Create an Intel82599 App for the device with 'pciaddress'.
 function Intel82599:new (arg)
    local conf = config.parse_app_arg(arg)
+   local self = {}
 
    if conf.vmdq then
       if devices[conf.pciaddr] == nil then
-         devices[conf.pciaddr] = {pf=intel10g.new_pf(conf):open(), vflist={}}
+         local pf = intel10g.new_pf(conf):open()
+         devices[conf.pciaddr] = { pf = pf,
+                                   vflist = {},
+                                   stats = { s = pf.s, r = pf.r, qs = pf.qs } }
       end
       local dev = devices[conf.pciaddr]
       local poolnum = firsthole(dev.vflist)-1
       local vf = dev.pf:new_vf(poolnum)
       dev.vflist[poolnum+1] = vf
-      return setmetatable({dev=vf:open(conf)}, Intel82599)
+      self.dev = vf:open(conf)
+      self.stats = devices[conf.pciaddr].stats
    else
-      local dev = intel10g.new_sf(conf):open()
-      if not dev then return null end
-      return setmetatable({dev=dev, zone="intel"}, Intel82599)
+      self.dev = assert(intel10g.new_sf(conf):open(), "Can not open device.")
+      self.stats = { s = self.dev.s, r = self.dev.r, qs = self.dev.qs }
+      self.zone = "intel"
    end
+   if not self.stats.counters then
+      self.stats.path = "/counters/"..conf.pciaddr.."/"
+      self.stats.sync_timer = lib.timer(0.001, 'repeating', engine.now)
+      self.stats.counters = {}
+      for _, name in ipairs(provided_counters) do
+         self.stats.counters[name] = counter.open(self.stats.path..name)
+      end
+      counter.set(self.stats.counters.type, 0x1000) -- Hardware interface
+      counter.set(self.stats.counters.dtime, C.get_unix_time())
+      counter.set(self.stats.counters.mtu, self.dev.mtu)
+      counter.set(self.stats.counters.speed, 10000000) -- 10 Gbits
```

Comment:
> Why is this a multiple of kilobits?

I guess this has sneaked in from my faulty SNMP code. For SNMP, there are two objects: `ifSpeed` and `ifHighSpeed`. The correct values for 10Gbps are `ifSpeed = 4294967295` and `ifHighSpeed = 10000` (RFC3635). My code used `ifSpeed = 10000000`, which is wrong and made it to here. The `speed` counter has nothing to do with SNMP and should say 10000000000 (hope I counted that right :).